notebooks/Bayesian Changepoint Detection.py

```diff
@@ -0,0 +1,234 @@
+
+# coding: utf-8
+
+# In[30]:
+
+get_ipython().magic(u'matplotlib inline')
+import numpy as np
+import scipy.stats as stats
+from matplotlib import pyplot as plt
+
+
+# In[31]:
+
+# Offline changepoint detection
+def changepoint_detection(data, hazard_function, distribution):
+    length = len(data)
+    H = hazard_function(np.arange(length + 1))
+    Pr = np.zeros((length + 1, length + 1))    # length + 1 b/c 0 is for priors
+    # Pr[r, t] = P(r_t=r, x_1:t)
+
+    Pr[0, 0] = 1
+    for t, x in enumerate(data):
+        t += 1    # t=0 is for priors stuff. First data point is @ t=1
+        predprobs = distribution.pdf(x)
+        Pr[1:t + 1, t] = Pr[:t, t - 1] * predprobs * (1 - H[:t])
+        Pr[0, t] = np.sum(Pr[:t, t - 1] * predprobs * H[:t])
+        Pr[:, t] /= np.sum(Pr[:, t])    # normalize probabilities
+        distribution.update(x)
+    return Pr
+
+
+# In[32]:
+
+class BayesianChangepointDetection(object):
+    def __init__(self, hazard, distribution):
+        self.hazard = hazard
+        self.distribution = distribution
+        self.time = 0
+    
+        self.Pr = np.ones(1)
+    
+    def step(self, x):
+        self.time += 1
+
+        H = self.hazard(np.arange(self.time))
+        old = self.Pr
+        self.Pr = np.zeros(self.time + 1)   # +1 because 0 <= r <= time
+        
+        predprob = self.distribution.pdf(x)
+        self.distribution.update(x)
+        
+        self.Pr[1:] = old * predprob * (1 - H)
+        self.Pr[0] = np.sum(old * predprob * H)
+        self.Pr /= np.sum(self.Pr)
+        
+        return self.Pr
+
+
+# In[33]:
+
+class Distribution(object):
```

Comment:
What is the inheritance doing for us here? Is there any reason to use it other than type-checking and self-documentation?